src/components/Documentation/Markdown/index.js

```diff
@@ -1,279 +1,145 @@
-  }
-
-  componentWillUnmount() {
-    document.removeEventListener('touchstart', this.onTouchStart)
-    document.removeEventListener('touchend', this.onTouchEnd)
-  }
-
-  isInsideCodeBlock = elem => {
-    for (let el = elem; el && el !== document; el = el.parentNode) {
-      if (el.tagName === 'PRE') return true
-      if (el.tagName === 'ARTICLE') return false
+const renderAst = new rehypeReact({
+  createElement: React.createElement,
+  components: { details: Details, abbr: ABBR }
+}).Compiler
+
+export default function Markdown({
+  htmlAst,
+  prev,
+  next,
+  tutorials,
+  githubLink
+}) {
+  const touchstartXRef = useRef(0)
+  const touchendXRef = useRef(0)
+  const isCodeBlockRef = useRef(false)
+
+  const handleSwipeGesture = useCallback(() => {
+    if (isCodeBlockRef.current) return
+
+    if (touchstartXRef.current - touchendXRef.current > 100) {
+      navigate(next)
     }
-    return false
-  }
 
-  onTouchStart = e => {
-    this.isCodeBlock = this.isInsideCodeBlock(e.target)
-    this.touchstartX = event.changedTouches[0].screenX
-  }
+    if (touchendXRef.current - touchstartXRef.current > 100) {
+      navigate(prev)
+    }
+  }, [prev, next])
 
-  onTouchEnd = () => {
-    this.touchendX = event.changedTouches[0].screenX
-    this.handleSwipeGesture()
-  }
+  const onTouchStart = useCallback(e => {
+    isCodeBlockRef.current = isInsideCodeBlock(e.target)
+    touchstartXRef.current = event.changedTouches[0].screenX
+  }, [])
 
-  handleSwipeGesture = () => {
-    if (this.isCodeBlock) return
-    const { prev, next } = this.props
+  const onTouchEnd = useCallback(() => {
+    touchendXRef.current = event.changedTouches[0].screenX
+    handleSwipeGesture()
+  }, [])
 
-    if (this.touchstartX - this.touchendX > 100) {
-      Router.push({ asPath: PAGE_DOC, pathname: next })
-    }
+  useEffect(() => {
+    document.addEventListener('touchstart', onTouchStart, false)
+    document.addEventListener('touchend', onTouchEnd, false)
 
-    if (this.touchendX - this.touchstartX > 100) {
-      Router.push({ asPath: PAGE_DOC, pathname: prev })
+    return () => {
+      document.removeEventListener('touchstart', onTouchStart)
+      document.removeEventListener('touchend', onTouchEnd)
     }
-  }
+  }, [])
 
-  render() {
-    const { markdown, githubLink, prev, next, tutorials } = this.props
-
-    return (
-      <Content id="markdown-root">
-        {tutorials && (
-          <TutorialsWrapper>
-            <Tutorials tutorials={tutorials} compact={true} />
-          </TutorialsWrapper>
+  return (
+    <Content id="markdown-root">
+      {tutorials && (
+        <TutorialsWrapper>
+          <Tutorials tutorials={tutorials} compact={true} />
+        </TutorialsWrapper>
+      )}
+      <GithubLink href={githubLink} target="_blank" rel="noreferrer noopener">
+        <i /> Edit on GitHub
+      </GithubLink>
+      <div className="markdown-body">{renderAst(htmlAst)}</div>
```

Comment:
so, what is the plugin that holding us from the danger html?

Reply:
Basically those two blocks: https://github.com/iterative/dvc.org/blob/5f589deee8fa32aca0f8c498947649fa02a87099/src/components/Documentation/Markdown/index.js#L31-L57

Reply:
it should be straightforward similar to what we did already on the blog side a few times?

Reply:
It's a little harder here. We just replaced one html with another in the blog. But for example `<Tooltip>` is a custom react component with a lot of custom logic for positioning. And `<Collapsible>` is using amitations and react state to show and hide content. We can't just replace them with static html.

Reply:
that's sad ... mostly means JS bundle will be way bigger? did you compare it already? Also means probably that we spent quite a lot of time on the blog side for no reason - they will be using the same JS, right?

Reply:
Not exaclty. We remove around 90% of code and libraries that we have before here: `react-markdown` (21kb minified), `react-syntax-hightlighter` (hard to say exactly, all themes and code is 397kb minified, but we probably used only 50-150kb in our case), `unist-util-visit` and our custom plugins for links and custom syntaxes. It's a lot of code.

We added new library `rehype-react` that have weight of `7.7kb` gzipped which is not perfect, but bearable and increased size of page data json because we are now sending not html, but parsed html ast. But these changes again are partially mitigated by gzip because ast is easy to compress. So in the end we have more code than we could if we just use static html, but significantly less then before migration.

Also blog pages will not be affected by it. Gatsby and next both create separate js-chunks for different page types, so `rehype-react` will only be inside doc page chunk.

Reply:
yeah, I'm just wondering would have it bern then easier on the blog end to the same with plugins - we spent quite a lot of time writing them ... or am I missing something? at least I know this for the future decisions ...

Reply:
Hard to say for sure. Simple tag replacement can probably be made simpler in browser but would increase bundle size. But stuff like that we did for image sizes are better implemented on the server.